synapse/handlers/stats.py

```diff
@@ -52,4 +59,324 @@ def __init__(self, hs):
     def notify_new_event(self):
         """Called when there may be more deltas to process
         """
-        pass
+        if not self.hs.config.stats_enabled:
+            return
+
+        lock = self.store.stats_delta_processing_lock
+
+        @defer.inlineCallbacks
+        def process():
+            try:
+                yield self._unsafe_process()
+            finally:
+                lock.release()
+
+        if lock.acquire(blocking=False):
+            # we only want to run this process one-at-a-time,
+            # and also, if the initial background updater wants us to keep out,
+            # we should respect that.
+            try:
+                run_as_background_process("stats.notify_new_event", process)
+            except:  # noqa: E722 – re-raised so fine
```

Comment:
You probably want a `finally` block here?

Reply:
That's not the case, we only want to release the lock if we can't start the background process

Reply:
Oh, I see what's happened here. In general its good to follow the rule of "one release per acquire". I think in your case you may want to do the acquire in the `process` function (and have a `if not lock.locked` guard outside)